### PR TITLE
Added initial work for syntax highlighting

### PR DESCRIPTION
**Pull request by bkonyi**
This change introduces a TextMate grammar based parser which outputs scoped span information corresponding to portions of the source which are part of the grammar. These ScopeSpans will be used to perform syntax highlighting in the code view.

### CODE REVIEW

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+  static List<_Matcher> _parseMatchers(List<Map<String, dynamic>> patterns) {
+    final List<_Matcher> parsed = [];
+
+    void _parseMatchersHelper(List<_Matcher> patternList) {
+      if (patternList == null) {
+        return;
+      }
+      for (final pattern in patternList) {
+        if (pattern is _MultilineMatcher) {
+          _parseMatchersHelper(pattern.patterns);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:71]: should this be `_parseMatchers`?

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+      'scopeName': scopeName,
+      'patterns': topLevelMatchers.map((e) => e.toJson()).toList(),
+      'repository': repository.toJson(),
+    });
+  }
+
+  final String name;
+  final String scopeName;
+  final List<_Matcher> topLevelMatchers;
+  final Repository repository;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:99]: move final vars below constructor or below static methods

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+        _start = start,
+        _end = end;
+
+  int get length => _end - _start;
+
+  final int _start;
+  int _end;
+
+  final int line;
+  final int column;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:118]: new line between members

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+    }
+    final result = callback();
+
+    if (matcher.name != null) {
+      _scopeStack.removeLast();
+    }
+    return result;
+  }
+
+  static final ListQueue<String> _scopeStack = ListQueue<String>();
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:142]: nit: style - move static methods above non static methods

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+
+  @override
+  String toString() {
+    return '[$_start, $_end, $line:$column (len: $length)] = $scopes';
+  }
+
+  /// Adds [matcher.name] to the scope stack, if non-null. This scope will be
+  /// included in each [ScopeSpan] created within [callback].
+  static List<ScopeSpan> applyScope(
+      _Matcher matcher, List<ScopeSpan> Function() callback) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:130]: trailing comma

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+  }
+
+  Map<String, dynamic> toJson() {
+    return {
+      for (final entry in patterns.entries)
+        entry.key: entry.value.map((e) => e.toJson()).toList(),
+    };
+  }
+
+  final Map<String, List<_Matcher>> patterns = {};
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:169]: nit: style, move class members above methods

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+        line: line + 1,
+        column: column + 1,
+      ));
+    }
+    return spans;
+  }
+
+  Map<String, dynamic> toJson();
+
+  final String name;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:243]: nit: ditto here. move above class methods

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+  Map<String, dynamic> toJson() {
+    return {
+      if (name != null) 'name': name,
+      'match': match.pattern,
+      if (captures != null) 'captures': captures,
+    };
+  }
+
+  final RegExp match;
+  final Map<String, dynamic> captures;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:281]: move final members above methods

File: packages/devtools_app/lib/src/debugger/span_parser.dart (hunk 1 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/debugger/span_parser.dart:71]: Actually, it looks like this is just old code. I'll remove it.

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 9]
```diff
@@ -159,14 +145,14 @@ class Repository {
     }
   }
 
+  final Map<String, List<_Matcher>> patterns = {};
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:148]: move type to RHS

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+    if (!scanner.scan(end)) {
+      return null;
+    }
+    if (contentName == null || (endCaptures ?? captures) != null) {
+      return _applyCapture(scanner, endCaptures ?? captures, line, column);
+    } else {
+      // If there's no explicit captures and contentName is provided, don't
+      // create a span for the match.
+      return [];
+    }
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:411]: this code is duplicated in `_scanBegin`. Let's put into a helper.

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+            scope: contentName,
+            column: column,
+            line: line,
+            start: start,
+            end: scanner.position,
+          ));
+        } else {
+          results.addAll(_scanUpToEndMatch(grammar, scanner));
+        }
+        final List<ScopeSpan> endSpans = _scanEnd(scanner);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:451]: remove type from LHS

File: packages/devtools_app/lib/src/debugger/span_parser.dart [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+class _IncludeMatcher extends _Matcher {
+  _IncludeMatcher(String include)
+      : include = include.substring(1),
+        super._({});
+
+  static bool isType(Map<String, dynamic> json) {
+    return json.containsKey('include');
+  }
+
+  final String include;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/span_parser.dart:512]: move below constructor

**Review by kenzieschmoll — APPROVED** [review]

**Pull request merged**